### ISSUE TITLE
Visualize, plot OPF experiment results

### ISSUE THREAD

**Issue by breznak** [numenta/nupic#2658]
I'd like to visualize the results `.csv` from running OPF experiment. 
It should be relatively easy, but perhaps some of you have already written a nice utility for that? 
NAB, @rhyolight ?

------------------------------
WORKING branch:
https://github.com/breznak/nupic/tree/plot_results

TODO:
- [x] move or copy the script from [NAB](https://github.com/numenta/NAB/blob/master/nab_visualizer.html) to `NUPIC/scripts/visualization/`
- [x] Offline & Security
  - ~~plot.ly~~ publishes all data online, I think this is unacceptable for general use for us
  - "our script" is pretty nice, runs in browser on localhost
    - [x] I suggest downloading the rendering scripts for offline use
- [x] add example data for easy plotting test + doc 
 - [x] OPF results data - `opf_results/*.csv`
 - [x] NAB (small) data - `data_file* results_file* data/ results/`
- [x] eval which plotting lib to use, currently it's `DyGraph` https://github.com/numenta/NAB/blob/master/nab_visualizer.html#L4-L7 -- @jefffohl liked `D3.js` for being more flexible, but decided to stick with `DyGraph` so far as it fits the current needs of plotting graphs.
- [x] extend to plot OPF data, not only NAB (subset of OPF) data (Help wanted, I can explain what to do with OPF data, provide parser script)
  - [x] plot single file <br/>
The current JS code from NAB spends a lot of work in parsing a structure of files, in order to plot all *.csv files there. For OPF/Nupic we don't need that (although it wouldn't hurt to keep the functionality for NAB). <br/>
What we need is the ability to `plot a single file provided as an argument` (`python plot.py ./opf_results/DefaultTask.csv`)
  - [x] merge data and results <br/>
NAB uses separate folders for paths & results, on OPF results file, the all these are together  
- [ ] improve the Plot page interface (Help wanted)
  - [x] plot data, add a checkbox option, the OPF field is `actual`
  - [x] plot anomaly, add a checkbox option, the OPF field is `anomalyScore`
  - [x] plot predicted field, add a checkbox option, the OPF field is `multistepBestPrediction.1`
    - [x]  optional, add option to plot other multistep (>1) prediction fields
    - [x]  opt, add textfield to enter specific label that should be plotted, maybe can fit ^^^
     - checkboxes generated dynamically for all CSV labels
     - [ ] preselect some values -- impossible to preselect/know data - but anomalyScore & multiStepBestPredictions.actual make sense 
    - [ ]  opt, plot labeled anomalies (not present in OPF, could be just a specific field)
  - [x] UI: merge data & results windows? 
  - [x] UI: anomalyScore [0..1.0] is not noticable with "raw data" plotted, FIX by rescaling (to say 90% of max of the data)? 
  - [ ] UI: smooth zoom out. Now can soothly zoom in be selecting the are of interest. and zoom out by mouse-click to the original size. Can we zoom-out iteratively as well (mouse wheel/a scroll-bar, ...)?  
  - [ ] UI: add highlight menu: check-boxes fields (ex. anomalyScore), threshold (0.9), (optionally: below/over), (opt.: color) and highlight the section where the field is over the threshold. 
   - [ ] Consider an AND operator for the 2 above statements, can be used as evaluation as follows: see https://github.com/breznak/nupic/pull/16<br/> 
(anomalyScore, >0.9) AND (annotation, >0.9), green # correctly detected anomaly<br/>
(anomalyScore, <=0.9) AND (annotation, >0.9), red # missed<br/>
(anomalyScore, >0.9) AND (annotation, <=0.9), yellow # false positive<br/>
- [ ] UI: On higher zoom-out levels (~1 week) it is not possible to see time (fine-grained) on x-axis interactively. (I think best solution would be if the `timestamp` field would show up in the "fields div on the left", as interactive values are shown there.) see https://github.com/breznak/nupic/pull/20
- [x] opt, FIX plotting of NAB Results, currently only data seem to work. (Help wanted) <br/>
  If the above succeeds, maybe NAB should switch to using OPF format for the output (data+results) ( @subutai ?)
- [ ] Extending OPF: issues not directly for this PR, related to OPF; @rhyolight ?
 - [ ] `anomalyScore` has a string type in OPF (maybe bcs it's None at the 1st step)
 - [ ] new `anomalyAnnotations` OPF field, to mark human annotations, could be useful in NAB
 - [ ] Plotting non-numeric inputs. Currently impossible, but each encoder has a `scalarValue` member, can we expose that (add a field `XX.scalar` for each input field to OPF file) somehow? 
- [x] Bugs
 - [x] selecting a non-OPF csv crashes the web app, can't render after reselecting a correct OPF file (w/o server restart) @jefffohl 
 - [x] fields with too long name (eg metrics) get shortened in the UI, but it's not possible to see the whole name. Maybe a "context label" on mouse hover? 
 - [x] "generic CSV" support (NAB,..) <br/> currently some fields are hard-coded (eg for "scaled anomaly score") and the rendering fails if the fields are not present. But a nice feature is the dynamic menu for plotting numeric fields, it would be nice if only the "scaled" function failed (and its checkbox is greyed-out) instead of all plot failing. 
   - [x] additionally, if the "scaled" functionality could be written more generic (an array for labels to scale) so we could scale both anomaly & likelihood score, or spiketrain (0/1) data,... ?
 - [x] /low priority/ Does not render in FireFox (nor did the orginal NAB code, but the DyGraph examples work fine in a recent FF) 
- [ ] New Features
 - [ ] Online plotting <br/> would require: a) the model sends the updates to the server; b) a network in/out region; c) refreshing the plot of the (updated) file every second or so.. 
- [ ] PR ready code
 - [ ] finish main Readme, add images, maybe a wiki
  - [ ] mention the "2 rows skipped" feature/problem
 - [x] add comments/docs to functions
 - [x] avoid hard-coded values, allow "settings" at code-level
 - [ ] should we add test-case for this code? 

----------------
UPDATES:
- 4/11/2015 - some refactoring, configurable options, improved "generic CSV" handling, documentation
- 3/11/2015 - Merged Jeff's PR with Bugfixes & overhauled usability! :clap: "Zoom view", and much more!
- 27/10/2015 - Merged Jeff's work enabling OPF files plotting, improving UI
- 24/10/2015 - Updated with @jefffohl 's "Initial commit" work, plots NAB data & results, improved checkboxes for plots

**Comment by rhyolight** [comment]
All my tools have been in matplotlib, which makes me want to die. But this could be updated to do even more generic plotting? https://github.com/numenta/nupic/blob/master/examples/opf/clients/hotgym/anomaly/one_gym/nupic_anomaly_output.py

**Comment by breznak** [comment]
Thanks, I'll check your script soon. 

Meanwhile I've tested the `in browser` plotting on NAB and it looks awesome :+1: Maybe we could bring in over to nupic and just edit it to parse OPF result files ? 

NAB has also a plotter using https://plot.ly/ which I haven't tested yet (need to set up an API key)

**Comment by breznak** [comment]
CC @subutai @BoltzmannBrain 
@dundalek Hey! It's me, Mark :wink: Can you suggest a nice (and simple) JS lib for visualizing CSV graph data?

**Comment by rhyolight** [comment]
I'm a big fan of http://dygraphs.com/. That's what River View uses.

---------
Matt Taylor
OS Community Flag-Bearer
Numenta

**Comment by BoltzmannBrain** [comment]
@breznak plotly has some great functionality for creating visually-appealing graphs, and for free. We use it in NAB [here](https://github.com/numenta/NAB/blob/master/scripts/plot.py). [Here's an example](https://plot.ly/~alavin/3137/anomaly-detections-for-twitter-data-from-goog/) from that script, where the green and red diamonds mark TP and FP detections, respectively, and the red dots are the true anomaly labels.

**Comment by breznak** [comment]
@rhyolight @BoltzmannBrain Yes, I've edited the issue with both. DyGraphs are used in 1 approach. The plot.ly graph looks really pretty, but the fact it publishes all the data you plot (unless you pay) is a blocker for some of our purposes - ei. we can't publish the data.

**Comment by rhyolight** [comment]
So what secret supervillain project are you working on?

---------
Matt Taylor
OS Community Flag-Bearer
Numenta

**Comment by breznak** [comment]
> So what secret supervillain project are you working on?

Nah, Skynet and stuff. :wink: And some medical records..

**Comment by breznak** [comment]
I've updated my initial steps in the working branch https://github.com/breznak/nupic/tree/plot_results 
I'm looking for a `JS/DyGraph` wizard who would be so kind in helping me with this issue :sos: as I think it'll be helpful to have for the upcoming HTM Challenge. ( @rhyolight @jefffohl ?)

**Comment by jefffohl** [comment]
I haven't tinkered with DyGraph before, but it sounds like the kind of thing that is up my alley. I might be able to clear out some time later this week to take a look.

**Comment by breznak** [comment]
Jeff, that would be awesome!
I'll prepare today some example data so it's in a runable form. Thanks a
bunch

**Comment by jefffohl** [comment]
@breznak - will this be used in your HTM Challenge submission? Or is it simply an enhancement to NuPIC that everyone can use? If the former, I probably shouldn't help out, unfortunately, because I am on the HTM Challenge judging panel, and it would create a conflict of interest.

@rhyolight - thoughts here?

**Comment by rhyolight** [comment]
@jefffohl If you are doing work that contributes to the NuPIC codebase, that is fine. Just don't help out by contributing to anyone's private project repos.

**Comment by jefffohl** [comment]
OK - thanks @rhyolight

**Comment by breznak** [comment]
> @breznak <https://github.com/breznak> - will this be used in your HTM
Challenge submission? Or is it simply an enhancement to NuPIC that everyone
can use? If the former, I probably shouldn't help out, unfortunately,
because I am on the HTM Challenge judging panel, and it would create a
conflict of interest.

@jefffohl No worries, I haven't joined a Challenge project yet. This is
intended for general Nupic enhancement, but I'd love to make it asap before
the challenge as I see it could be useful to some people during their work
on HTM challenge projects.

**Comment by jefffohl** [comment]
@breznak - OK, sounds great. Just trying to be careful about my responsibilities.

**Comment by breznak** [comment]
I've updated the TODO steps and ordered them sequentially

**Comment by subutai** [comment]
> Meanwhile I've tested the in browser plotting on NAB and it looks awesome  Maybe we could bring in over to nupic and just edit it to parse OPF result files ?

No one is really maintaining this right now. It does look it doesn't plot the results files correctly anymore. It would be nice if someone could fix that.  The UI is rather cumbersome too. It does work fine for examining the raw data files.

**Comment by jefffohl** [comment]
Working on this now. First, I am evaluating some different plotting libraries. Let me know if anyone is tied to DyGraphs.

**Comment by jefffohl** [comment]
@rhyolight - when you were choosing a data visualization tool for River View, did you take a look at [D3.js](http://d3js.org/)?

I am kind of leaning towards D3.js because it is more of a low-level framework, and therefore is more flexible. It occurs to me that, although we are solving a specific problem here, we are likely to want to continue to build various ways to visualize various things - therefore having a framework (rather than a tool) might be a better choice.

Does anyone have strong opinions here?

**Comment by breznak** [comment]
@jefffohl no strong opinions, as he who writes the code decides :wink: And both frameworks look really pretty to my amateur eye. But I don't see what else than OPF result graphs would we want to visualize in nupic? Remember there is the NuStudio to show you graphical representations of HTM and its parts..

**Comment by rhyolight** [comment]
I have using D3 directly in the past and it is an excellent library.

**Comment by breznak** [comment]
@jefffohl I'll probably start looking into this too, can you please share if you had some progress? And if/why do you see D3.js better than DyGraph?

**Comment by jefffohl** [comment]
@breznak - sorry about my slow progress - I have to squeeze this work in with my day job :). 

I was considering D3.js only because it is more of a framework, so that might serve us better in the future, if we intend to do more. But, for now, it seems that DyGraph will probably be the easiest to work with, so that seems the best bet at the moment. If we want to get more fancy later, we can.

I haven't done much except for reviewing charting frameworks, and getting up to speed with NAB.

I was going to spend some time today working on this, but if you want to divide up tasks, we could do that too.

**Comment by breznak** [comment]
Thanks Jeff, no worries. I just wanted to use your experience to know which
framework I should start learning about..so dygraph it is for now.
I will be learning my ways around, and if I started doing some real work,
I'll sync with you here so we don't duplicate.

PS: can I help something with NAB? But I think we don't need it here..(?)

**Comment by jefffohl** [comment]
@breznak - ok. I am going to keep at it, so let me know if you start to work on any of the issues on your checklist.

Regarding NAB - I just wanted to see how NAB fit into OPF - so not intending to use it, just wanted to understand better where you are coming from in terms of what your needs are.

**Comment by jefffohl** [comment]
@breznak I am making some progress - I might have something to share tomorrow.

**Comment by jefffohl** [comment]
HI @breznak. I have put together something. Right now, it is an interface that shows two panes: data and results. For each, there is a select menu that allows the user to select which CSV file they want to plot.

Is this what you are imagining, or are you thinking that the data should be merged into one graph?

I can make a pull request if you want to see the work in progress.

**Comment by jefffohl** [comment]
Here is a screen shot for reference: ![screen shot 2015-10-23 at 1 35 23 pm](https://cloud.githubusercontent.com/assets/641147/10704361/2756ce2c-798b-11e5-975d-4173d9f24c5c.png)

**Comment by breznak** [comment]
Hi @jefffohl , it looks really nice! :+1: :smile: Please make a PR, or tell me your branch, I'd love to test it out.

**Comment by cogmission** [comment]
Agreed, it DOES look nice!



On Fri, Oct 23, 2015 at 4:57 PM, breznak <notifications@github.com> wrote:

> Hi @jefffohl <https://github.com/jefffohl> , it looks really nice! [image:
> :+1:] [image: :smile:] Please make a PR, or tell me your branch, I'd love
> to test it out.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numenta/nupic/issues/2658#issuecomment-150702560>.
>



-- 
*With kind regards,*

David Ray
Java Solutions Architect

*Cortical.io <http://cortical.io/>*
Sponsor of:  HTM.java <https://github.com/numenta/htm.java>

d.ray@cortical.io
http://cortical.io

**Comment by jefffohl** [comment]
Here you go: https://github.com/breznak/nupic/pull/9

**Comment by breznak** [comment]
A couple of questions/ideas to discuss:
- about the "separate data", what do you think? With the nice check-box options for what to plot, I see little importance in having `data/results` in separate windows. 
- the checkbox functionality for labels is really awesome! Is it hard-coded, or works for any CSV?
- now it accepts the [NAB data format](https://github.com/breznak/nupic/blob/plot_results/scripts/visualization/results/numenta/realKnownCause/numenta_nyc_taxi.csv), for NuPIC we need to focus on the [OPF format](https://github.com/breznak/nupic/blob/plot_results/scripts/visualization/opf_results/DefaultTask.TemporalAnomaly.predictionLog.csv) too. 
 - will the CSV parsing library handle a cell  `{with, coma-separated, fields, inside brackets}` ? <br/> I have a bash script to handle this, which I will try to convert to python if it doesn't.
 - the OPF CSV typically have a lot of labels, should we filter them (or hardcode some "regexp" to filter them?)
- can I zoom in/out the graph, please? :wink: 
- only for NAB (or maybe later NuPIC), should we focus on plotting `annotated anomaly labels`?

**Comment by jefffohl** [comment]
* I can code it up to merge the data and results data sets. The only question I have is, how to make sure we are loading the proper files? Will the paths to the data and the results always have reliable relative paths? Or, should we let the user choose a data file (based on a manifest file as we have now), then a results file, then merge them together after they have been selected?
* the checkboxes are dynamically generated based on the CSV header. so, they will work with any CSV.
*  The DyGraph CSV parser choked on the OPF format. I can write a Javascript parser for that. I am assuming that we want it to strip out all fields that are not a number? Since the second line indicates the data type, this should be relatively easy. Or do we want to somehow convert some of the string values to numbers in some way? Note that DyGraph can only plot numbers (integers or floats).
* You can zoom into the graph by using your cursor to select a section of the graph. Though, I just noticed that I introduced a bug here. Normally, double-clicking will return the graph to the zoomed-out state. But I accidentally disabled that with some code that displays the timestamp when you click on the graph. I will fix that.
* Unfortunately, I am ignorant about annotated anomaly labels. Can you elaborate on that point?

**Comment by breznak** [comment]
btw @jefffohl I've added some UI questions (designated by "UI") to the issue description, esp. fixing the anomalyScore scale would be nice. Now back to replying to your comment...

**Comment by breznak** [comment]
> * I can code it up to merge the data and results data sets. The only question I have is, how to make sure we are loading the proper files? Will the paths to the data and the results always have reliable relative paths? Or, should we let the user choose a data file (based on a manifest file as we have now), then a results file, then merge them together after they have been selected?

The file lists are used in NAB which relies on the structure of the data, so I think it's safe to assume the paths will be correct. Actually NAB has an alternate plotting through `plot.ly` so I would not worry about that functionality too much. 

...if we focus on NuPIC/OPF, there are no separate data/results file lists. The results are in a single file (and raw data are included under the field `actual`). 
- How to pass the OPF file path to the script? 
 - can we use the command-line argument with http server? `visualization$ python -m SimpleHTTPServer 12345 ../examples/my/path.csv`  
 - can we trigger a file chooser dialog from the webUI? 
 - stick with the file list, which the user will manually generate and will contain paths to all CSVs that should be plotted. 
 - have the manifest-file with a folder and then show all CSV under that path?

**Comment by breznak** [comment]
> * the checkboxes are dynamically generated based on the CSV header. so, they will work with any CSV.

:guitar: :+1:

**Comment by breznak** [comment]
> *  The DyGraph CSV parser choked on the OPF format. I can write a Javascript parser for that. I am assuming that we want it to strip out all fields that are not a number? 

yes, but..

>Since the second line indicates the data type, this should be relatively easy. Or do we want to somehow convert some of the string values to numbers in some way? Note that DyGraph can only plot numbers (integers or floats).

Plotting only numbers is OK, the data-type line (2nd) would not work that easily, as for the example [OPF file here](https://raw.githubusercontent.com/breznak/nupic/plot_results/scripts/visualization/opf_results/DefaultTask.TemporalAnomaly.predictionLog.csv) the `anomalyScore` is type string: 
1. this is a NuPIC bug and we should ignore it and require it fixed ( @rhyolight ? I think that would be a correct approach)
2. hack around, my parser did well with replacing `{.*}` or `[]` to `0`;

PS: optionally, in the future we may want to get some of the other string fields, eg to get `multiStepBestPredictions.1,multiStepBestPredictions.5` (which are in fact again floats)

**Comment by breznak** [comment]
> * You can zoom into the graph by using your cursor to select a section of the graph. Though, I just noticed that I introduced a bug here. Normally, double-clicking will return the graph to the zoomed-out state. But I accidentally disabled that with some code that displays the timestamp when you click on the graph. I will fix that.

Cool thanks, I've noticed the zoom-out problem. Btw, do you think it would be possible to do some "smooth zoom-out" (the way zoom-in works)? With a mouse-wheel, a scroll-bar, ...?

**Comment by breznak** [comment]
> * Unfortunately, I am ignorant about annotated anomaly labels. Can you elaborate on that point?

This exists only in NAB so far *). Annotations is a TXT file with a vector, where `1` means a human-annotated anomaly = nice to plot with a "stem". Maybe another "Choose annotations file:" file input field would work fine here? A nice to have but definitely not needed right now. 

*) @rhyolight would it be any problem/benefit getting an `anomalyAnnotation` field to OPF? Useful here and then NAB could use OPF files, instead of its custom. I've raised that issue there but got no opinions..

**Comment by breznak** [comment]
@jefffohl please let me know if there's anything I can help, with a non-JS stuff.

**Comment by jefffohl** [comment]
* In determining how to let the user choose what OPF file to use, all of those sound like good suggestions. It depends mostly on how users typically use the OPF, and about this I am somewhat ignorant. The most universally easy solution would be to launch a file browser from the web UI. We could also allow users to put in a path manually, which could be a publicly accessible URL. So, maybe there would be a "path" field, where the user types in the path, and a button for launching a file system browser for finding and loading local files.
* For parsing OPF files, would it be OK to hard-code into our script exceptions for `anomalyScore`, `multiStepBestPredictions.1`, and `multiStepBestPredictions.5` and the like? That would make it pretty easy, as long as we can count on those fields always being number-like (we can use type coercion to turn them into numbers). It makes things more tightly coupled, but so far, this would only be for OPF, so that should be OK, right?
* For smooth zooming, it don't see an option for that in DyGraphs, but I will look around, and I may be able to add a plugin for that as well.
* Will worry about the anomalyAnnotation later. I think what I will end up having is two different versions of this script - though working in the same way - one for NAB, and one for OPF. Alternatively, we could perhaps use one script, but give it a config somehow, so it behaves slightly differently depending on if it is working with OPF data, or NAB data.

**Comment by jefffohl** [comment]
> @jefffohl please let me know if there's anything I can help, with a non-JS stuff.

Will do. Thanks!

**Comment by breznak** [comment]
> * In determining how to let the user choose what OPF file to use, all of those sound like good suggestions. It depends mostly on how users typically use the OPF, and about this I am somewhat ignorant. The most universally easy solution would be to launch a file browser from the web UI. We could also allow users to put in a path manually, which could be a publicly accessible URL. So, maybe there would be a "path" field, where the user types in the path, and a button for launching a file system browser for finding and loading local files.

I think a typical use case is a user running a single OPF experiment and wants to see the results. 
So a `a file browser from the web UI.` sounds like the way to go to me. 
Alternatively later we could add behavior like Gmail attachements, "Add a file" file-chooser and then checkboxes for which files to plot.

**Comment by breznak** [comment]
> * For parsing OPF files, would it be OK to hard-code into our script exceptions for `anomalyScore`, `multiStepBestPredictions.1`, and `multiStepBestPredictions.5` and the like? That would make it pretty easy, as long as we can count on those fields always being number-like (we can use type coercion to turn them into numbers). It makes things more tightly coupled, but so far, this would only be for OPF, so that should be OK, right?

Agree, simple & works. 
- `anomalyScore`, `multiStepBestPredictions.actual` are always present and a float. 
- `multiStepBestPredictions.1` is always present but `None` at the 1st row.
- `multiStepBestPredictions.N` - optional, can be more (3,5,10), but not very typical. None till N+1-th row. 

I still like your approach to plotting all numeric fields (as input data will be plotted if it is in a suitable form), are you planning combining the 2 approaches?

**Comment by breznak** [comment]
> * For smooth zooming, it don't see an option for that in DyGraphs, but I will look around, and I may be able to add a plugin for that as well.

(btw, just to make myself clear, I think the zoom-in with mouse selection is perfectly viable, just missing a zoom out step). 

I just found a mention of https://code.google.com/p/dygraphs/issues/detail?id=366 in Issue 58 some calls could exist (?)

**Comment by breznak** [comment]
> * Will worry about the anomalyAnnotation later. I think what I will end up having is two different versions of this script - though working in the same way - one for NAB, and one for OPF. 

100% agreed. Personally I'd try to push NAB to switch to using OPF, personally I don't see why not to.

**Comment by BoltzmannBrain** [comment]
@breznak could you please elaborate on what you mean by pushing NAB to use OPF?

**Comment by breznak** [comment]
> @breznak could you please elaborate on what you mean by pushing NAB to use OPF?

@BoltzmannBrain bad wording, sorry. I had an issue in NAB whether it could use OPF for its format (extended with a annotations column), so this code could be shared (eg the results plotting is fixed here)

**Comment by rhyolight** [comment]
Guys, this issues is just for plotting experiment results after they've been run, right? There is no initiative to visualize live predictions / anomalies coming out of NuPIC, is there?

**Comment by jefffohl** [comment]
Right now, the little web app I am working on just reads files that have already been created by the OPF. But, I was imagining that the next step would be to visualize streaming live data.

**Comment by rhyolight** [comment]
That's a good thing to think about, but it will require a server to post data to. The live data coming out of NuPIC will be from a python runtime, so it will need to be posted somewhere for display. Just something to think about.

**Comment by breznak** [comment]
> I was imagining that the next step would be to visualize streaming live
data.

Same. I was planning this for later. Actually if the plotting is reasonably
fast, you could just reread the file every second, to get a semi-live data
plot.

**Comment by breznak** [comment]
> ...so it will need to be posted somewhere for display.

In some other issue I was thinking about network I/O region (maybe somebody
already implemented that? I remember some networking code(?)) The OPF model
could just add a network-out region and the plotter would bind to a known
address.... could work like that? (but not there yet)

**Comment by jefffohl** [comment]
I created a new PR. Notes here: https://github.com/breznak/nupic/pull/10

Is it better to include the notes here, or with the PR?

**Comment by breznak** [comment]
> I created a new PR. Notes here: breznak#10

@jefffohl This is a HUGE progress, KUDOs! :+1: :100: This is everything what I've originally imagined for the visualizer. 

I've merged your work, updated the check-points and tested, all works well here. 

> I am not sure if this approach is optimal or not. It might be better to hard-code certain fields to be included. Or perhaps this way is best. Would like feedback on that. In any case, consider this a work in progress, with this being the latest installment as we work towards a more polished app.

I would be more for the dynamic approach with plotting all numeric fields, as we don't know what the user might be interested in (for example, I found quite interesting plotting all the metrics and comparing their performance). A compromise might be preselecting some default check-boxes (anomalyScore, bestPredictions.actual)?

**Comment by breznak** [comment]
@jefffohl I've added some usability comments to the "Bugs" section, what do you think about them?

**Comment by jefffohl** [comment]
Thanks @breznak. I will take a crack at the bugs today. As you are more familiar with the OPF than I am, can you give me a list of what fields will always be found in an OPF file, as well as fields that may or may not be found there? I would like to auto-detect the file type, to try and determine if it is an OPF file, and if it is a valid OPF file.

**Comment by jefffohl** [comment]
@breznak - also, what fields besides anomalyScore do we imagine a user would like to scale proportionally to the data range?

**Comment by breznak** [comment]
> can you give me a list of what fields will always be found in an OPF file, as well as fields that may or may not be found there? I would like to auto-detect the file type, to try and determine if it is an OPF file, and if it is a valid OPF file. 

@jefffohl this is little tricky. An OPF file will have columns by the `Model` used, if you take "anomaly detection" as the intended use-case for the visualizations, the model has to be `TemporalAnomaly` (which inherits from `TemporalMultiStep`) and always has an `anomalyScore` field. (+ the field(s) from TemporalMultiStep). 

A `TemporalMultiStep` will always have the `multiStepBestPredictions.actual` field, and depending on user's settings additional `multiStepBestPredictions.<N>` fields. 

There is also always the `timestamp`, but it may not have to be in a date format always, can appear as an iteration (1,2,3,4,...).

Other models are not actively used AFAIK. 

> what fields besides anomalyScore do we imagine a user would like to scale proportionally to the data range?

Do you intend this to be OPF-only, or a generic "I will plot what I can from the CSV" implementation? 
If the latter, one of the use-cases I had on mind was comparing several anomaly implementations/settings. This happens in NAB results file (it could be plotted if we overcome the hard-coded names somehow? *1) ) which includes the `raw_anomaly` and `anomaly_score` (=likelihood) fields. 

```
timestamp,value,anomaly_score,raw_score,label,S(t)_reward_low_FP_rate,S(t)_reward_low_FN_rate,S(t)_standard
2014-07-01 00:00:00,10844.0,0.0301029996659,1.0,0,0.0,0.0,0.0
2014-07-01 00:30:00,8127.0,0.0301029996659,1.0,0,0.0,0.0,0.0
2014-07-01 01:00:00,6210.0,0.0301029996659,1.0,0,0.0,0.0,0.0
2014-07-01 01:30:00,4656.0,0.0301029996659,1.0,0,0.0,0.0,0.0
``` 

*1) 2 ideas to approaching the "fixed names" problem: 
A) "in code":
```
NORMALIZE_TO_FIELD="multiStepBestPredictions.actual"
SCALE_FIELDS_ARR=["anomalyScore"] # can extend here
```
B) "in web GUI":
- a set of radio-buttons "Reference field for scaling..." 
- and check-boxes "Scale these fields"

**Comment by breznak** [comment]
Btw, I've found a working example for the finer zoom in/out 
http://dygraphs.com/gallery/#g/interaction ("Custom interaction" block)
but I'm still unable to transfer it here :/

**Comment by jefffohl** [comment]
Thanks @breznak.

I do want to make the app as flexible as possible, and therefore would like to enable it to consume whatever data is passed its way, with the only constraint being that there must be a `timestamp` field.

I like your idea for normalizing fields. The "fields" UI might start getting a bit crowded at this point, so we would probably want to allow the user to toggle this portion of the UI in and out of sight.

Regarding the zoom, feature, have you taken a look at the range-selector widget? http://dygraphs.com/gallery/#g/range-selector

**Comment by jefffohl** [comment]
@breznak do you have any ideas about how to render field names that are very long, such as this?
`multiStepBestPredictions:multiStep:errorMetric='aae':steps=5:window=1000:field=consumption`

Can you think of a more concise label that still has meaning for the user? If we can come up with some kind of logic, I can convert them.

**Comment by rhyolight** [comment]
@breznak @jefffohl I would like to test this out. How do I do it?

**Comment by jefffohl** [comment]
@rhyolight 
* Pull this branch
* `cd nupic/scripts/visualization`
* `python -m SimpleHTTPServer 8080`
* Use browser to view: `http://localhost:8080`

**Comment by jefffohl** [comment]
Oops - forgot this is on the main branch. Here is @breznak's branch: https://github.com/breznak/nupic/tree/plot_results

**Comment by jefffohl** [comment]
PS - you will want to test with an OPF output file, like the one found at nupic/scripts/visualization/opf_results/

**Comment by breznak** [comment]
> I do want to make the app as flexible as possible, and therefore would like to enable it to consume whatever data is passed its way, with the only constraint being that there must be a timestamp field.

Great, @jefffohl ! I hoped you'd say that :) Then I hope this project has a lot of potential users - NAB, RiverView, all the Nupic examples with custom plot implementations, ...

> I like your idea for normalizing fields. The "fields" UI might start getting a bit crowded at this point, so we would probably want to allow the user to toggle this portion of the UI in and out of sight.

That's true. You'd want to set the scaling up once and then continue browsing the graph. We could also already preselect some values (reference=bestPredictions.actual; scale=anomalyScore; ..if these fields are found. But not sure if this is not an overkill?) 

> Regarding the zoom, feature, have you taken a look at the range-selector widget? http://dygraphs.com/gallery/#g/range-selector

I haven't, and it looks pretty good and combines the functionality of un/zoom with a global outlook! :+1: 

> ...do you have any ideas about how to render field names that are very long, such as this?
`multiStepBestPredictions:multiStep:errorMetric='aae':steps=5:window=1000:field=consumption`
Can you think of a more concise label that still has meaning for the user? If we can come up with some kind of logic, I can convert them.

I think we can't compress it (much), if we go for some "encoding", it will become over complicated. 
We could:
A) drop the part `multiStepBestPredictions:multiStep` which is same for all metrics in our OPF model (but does not solve the general problem of long names)
B) can we use `<span>` tag on mouseover to get the full text and leave the normal text with overflow: hidden ? (imho preferred method) 
C) use some JS hackery, get the label's size, if too long, create a placeholder "Label <N>" and inject a paragraph 
```
<p>
Graph labels:
* label <N> = "blah blah"
```

**Comment by jefffohl** [comment]
@breznak - a new pull request for review. I made a lot of changes:

* The options panel to the right of the graph now shows the following:
 - Visibility toggle
 - Current highlighted value
 - Field name
 - Checkbox for normalizing data
 - Radio buttons for selecting the data field to which other fields can be normalized
* The normalization options can be hidden away to save space
* Added scrolling zoom mini-graph UI element at bottom of graph. Note that this is not compatible with the previous zoom method, wherein the user could draw an area on the graph to zoom into.
* Aesthetic improvements
* The app can now consume any CSV file, as long as it has a timestamp field. **However** because it is assuming the file to be an OPF file, it strips out the second and third lines of the CSV. This means that if a non-OPF file is consumed, it will work, but the first two steps of the series will be missing. So, we will have to figure out a way to determine if the file is produced by the OPF or not. Alternatively, this could be done manually through the UI, with a checkbox or something allowing the user to tell the app what kind of file it is.

**Comment by jefffohl** [comment]
A screen shot of what it should look like now: ![screen shot 2015-10-29 at 9 56 07 am](https://cloud.githubusercontent.com/assets/641147/10825725/8d88e148-7e23-11e5-9875-43e5479d1929.png)

**Comment by jefffohl** [comment]
I will be looking into the issue of DyGraph not working in Firefox.

**Comment by rhyolight** [comment]
Maybe it's not the right time for feedback... but I really like having a
smaller, synced graph to display the anomaly score and likelihood, if there
is one.

---------
Matt Taylor
OS Community Flag-Bearer
Numenta

**Comment by jefffohl** [comment]
@rhyolight - do you mean a smaller graph below the main graph? Can't the same be achieved with what we have now, where the user can turn on and off whichever fields they wish to view? Perhaps you can expand on your idea a bit more?

**Comment by rhyolight** [comment]
I should play with it first before you do anything. Will do that now.

---------
Matt Taylor
OS Community Flag-Bearer
Numenta

**Comment by jefffohl** [comment]
@breznak I added a fix for the Firefox issue to the current outstanding PR.

**Comment by rhyolight** [comment]
I am looking into how to use the [`BasicPredictionMetricsLogger `](https://github.com/numenta/nupic/blob/master/src/nupic/frameworks/opf/opfbasicenvironment.py#L168-168) with an OPF model outside the [`experiment_runner`](https://github.com/numenta/nupic/blob/master/src/nupic/frameworks/opf/experiment_runner.py) framework. I would ideally like to be able to plot the results of any model I create with `ModelFactory.createModel()`.

**Comment by jefffohl** [comment]
@rhyolight - Not being an expert with the OPF, can you explain to me what your comment implies? It looks like the BasicPredictionMetricsLogger outputs JSON. Are you saying that you would like the app to be able to consume any data format (and possible automatically figure out the format type)?

**Comment by rhyolight** [comment]
No, am just trying to figure out how to get the data file you are plotting out of the OPF. I don't use the OPF experiment framework for anything (which is what created that file), I always just create a model and run it.

**Comment by breznak** [comment]
> ..I always just create a model and run it. 

@rhyolight just write the outputs to a CSV file and you are good (?)

**Comment by rhyolight** [comment]
@breznak I don't see how I can do this in the OPF API.

**Comment by breznak** [comment]
Merged Jeff's latest PR and updated the issue description, the check-list is getting slimmer :wink: 
What do you think Jeff, are there any issues that should be ironed out, or can we head for a public PR to nupic soon? :fireworks:

**Comment by breznak** [comment]
> I don't see how I can do this in the OPF API. 

@rhyolight we're working towards reasonably generic CSV support (I'll have some testcase and new (sub)issues soonish), so basically anything with a `timestamp` field would work. The "main" focus on OPF files assumes the files produced by `scripts/run_opf_experiment` so you could look there how to write is a similar format (?) Otherwise it might be better to discuss under some specific code you're working on ?

**Comment by jefffohl** [comment]
@breznak I think this looks good, in terms of a big improvement over what is there now. We will need to write some tests before we can submit a PR, I believe, and this is something that we will need to work with @rhyolight on. I am not sure if there are any Javascript testing frameworks used inside of NuPIC, and whether or not we want to introduce them. Also, if we are going to do more work on this, I would like at some point to introduce some workflow tools, for building the app, such as gulp and SaSS, etc. I am not sure how Matt feels about having all of that stuff in NuPIC.

**Comment by rhyolight** [comment]
generic CSV is perfect for me.

---------
Matt Taylor
OS Community Flag-Bearer
Numenta

**Comment by jefffohl** [comment]
@breznak one thing that I mentioned, but which bears repeating, is that currently, the script will strip out the second and third rows of the CSV. So, if the file is not similar to the OPF output, the first two lines of data will be missing.

**Comment by breznak** [comment]
> one thing that I mentioned, but which bears repeating, is that currently, the script will strip out the second and third rows of the CSV. So, if the file is not similar to the OPF output, the first two lines of data will be missing.

@jefffohl ..I'm just writing a comment about related functionality, and I think dropping the first 2 lines is perfectly OK, as NuPIC is not trained there anyway. I'll put it in a readme, or a small 
-----------------------
*) notice on the page?

**Comment by jefffohl** [comment]
@breznak - sounds fine for now - I just wanted to make sure you were aware.

**Comment by breznak** [comment]
Merged a small example data cleanup - moved to `./examples/{OPF,NAB,CSV}` path. 

> generic CSV is perfect for me.

@rhyolight ...right on time! :hourglass: 

@jefffohl I'd like to discuss my user-experience on the latest iteration, esp. focusing at the "generic CSV": 

- The "range select" (sub graph at the bottom) works perfectly! However it replaced the "mouse-select a region to zoom", do you think it's possible to combine both of them in one graph? 

- generic CSV

I've added a testing/example file `examples/CSV/sin_30sec.csv` that looks like this:
```
time,function,anomaly_score
0.0,0.0,1.0
0.05,0.30901699,1.0
0.1,0.58778525,1.0
0.15,0.80901699,1.0
0.2,0.95105652,1.0
```

1/ missing `timestamp` field
I know we've stated a `timestamp` always occurs on an OPF file, but in cases when not (generic) we might... 
* Force it, "just rename your field to timestamp" (`time` -> `timestamp` in the example)
* Let the user select which field means time (axis X)
 * a column `Timestamp` with radio-buttons, done the same way `Data` is chosen for normalization. Probably the least disruptive change(?)
 * ~~~"Setup screen" - this would be a bigger overhaul to the UI (still using most of the existing functions). Thinking about it I don't like it as much as the idea above, but I'll include it here for the sake. The idea is to have a setup page where the user would do all the things he can do now: presented names of the columns, one would chose: which cols to plot (would simplify the automated guessing / but more difficult for an "out of the box" experience for some unknown (OPF)files), which field is timestamp, which to scale, ... Then he'd click "Next" and go to the graph.~~~
 * Default - iterations. Another approach would be assuming a default = generate a sequence `1..N` for the data items.

2/ non-OPF format (3 row header)
After workarounding 1/ by renaming to `timestamp`, I still couldn't draw the graph. The fix was to add the 2 more rows to be like an OPF header: 
```
time,function,anomaly_score
,,
,,
0.0,0.0,1.0
0.05,0.30901699,1.0
...
```
I'm not sure why as the code should just skip the 2nd and 3rd rows. 

3/ non DateTime timestamp
Another problem was my time (or axis X) was not in a date format but a sequence from 0.0 with 0.05 delta. The graph "ploted" but the range selector had troubles with the numbers, so the graph looks empty. 

It would be nice if we could support this more generic "time" format. Especially as it would make it possible for the default `1..N` if no timestamp is given. I'm not sure how this collides with the range selector (is it tied to datetime format?)


What is your opinion about these?

**Comment by jefffohl** [comment]
@breznak point by point:

* The range-select is incompatible with the mouse-zoom feature. It is either one or the other (unless we want to see if we can make a change to DyGraphs - which I don't want to try right now). Personally, I prefer the mouse-zoom feature.
* For dealing with CSV files that might have a different name than "timestamp", for now, I suggest we leave it, requiring the user to have a field named, "timestamp". In a later iteration, we can pop up a modal that says something like, "We noticed that your data has no 'timestamp' field. Would you like to select the field that represents the time data?" and show a list of the fields. After doing this, the user can then click the "Render" button.
* I will look into the error that occurs on CSV files that do have timestamps, but aren't able to be rendered.
*  DyGraphs requires that all rows have a valid Javascript date object. In Javascript, we create a Date object in this format: `new Date(year, month[, day[, hour[, minutes[, seconds[, milliseconds]]]]]);` So, if we pass in a float, it won't really know what to do with that, as that would indicate a floating point year. This means we will have to write some logic that will handle a variety of submissions. I am not sure how to do this other than to write rules for specific types of timestamps (integers, floats, etc.). Alternatively, we could make it accept *only* ISO 8601 formatted strings (which would be the sane thing to do, but isn't compatible with the examples you have given me). Right now, I have some logic in there that will accept dates in the following formats:
 * yyyy:mm:dd hh:mm:ss:ms
 * mm/dd/yyyy hh:mm:ss:ms (American style)

**Comment by jefffohl** [comment]
@breznak - I figured out what was going wrong with your test file. The script is assuming that all timestamps are in string format, and it is choking when it encounters a number. It seems to me that it will be difficult to allow timestamps to be in any format the user wishes. Ideally, it would have to be in the ISO 8601 format, but we could define some limited list of valid formats, and warn the user about this requirement through the UI and the README file.

**Comment by breznak** [comment]
Hi @jefffohl ! 

* for "understanding the graph data", I would prefer the mouse-zoom too, sorry to be a pain.. 
* `timestamp` column - let's just enforce it then. 
* about DyGraphs & Date: 
Sorry if I'm ignorant, but does it really have to be a date format for 1 column? I found http://dygraphs.com/data.html#array and it seems logical it's a matrix with numeric columns. 
Could we do in `generateFieldMap()` something like: 
```
if timestamp is numeric: 
 // keep timestamp column as is
eleif timestamp is Date:
 timestamp = ParseDate(timestamp) 
else: // no column named timestamp
 timestamp = range(1:1:size(data)) // create dummy default
```
...It seems it really is possible to do the dates/numbers for x-values, http://dygraphs.com/data.html#csv , now I probably don't understand the use for `convertPapaToDygraph()` function, which is dealing with dates and hardcodes the Date conversion? As DyGraph should handle the date format (is this what you meant by expecting the ISO dates?) Can we force ISO/numeric timestamp and let DyGraph decide to handle both seamlessly?

**Comment by breznak** [comment]
added a few "PR ready code" TODOs, working on a simple code "cleanup"

**Comment by breznak** [comment]
@jefffohl please sync with me now if you started working on something, I'm finishing some changes, so that we don't duplicate..

**Comment by breznak** [comment]
... and finished what I've meant in https://github.com/breznak/nupic/pull/13

**Comment by breznak** [comment]
Last thing I'd love to see is the "conditional highlighting" (2 TODOs in UI)

**Comment by jefffohl** [comment]
@breznak I was mistaken that DyGraphs requires a Date object. It will accept a number or a Date object when the data is submitted as an array. We want to use an array, because that is easier to work with, from what is output from Papa Parse.

So, if the timestamp is of type Number, we can just pass that along without needing to create a Date object.

Note that the reason we are creating our Date objects the way that we are (as opposed to just passing a string to the Date constructor), is that Firefox is more strict about what formats it will accept. So, some user agents will be more accommodating than others. Therefore, we have to be precise about how construct the Date object.

Thanks for improving the opf_visualizer.js script. I am making a branch for myself also called `plot_results_refactor` and I will fix some bugs and make some more improvements.

**Comment by breznak** [comment]
Thanks @jefffohl ! 
Btw, have you reviewed breznak#13 and agree with it? If it doesn't collide with any of your unmerged code, I'd like to merge it to avoid future conflicts.

**Comment by jefffohl** [comment]
@breznak I haven't written anything since you created `plot_results_refactor`, so you can merge it if you like. There are some changes I would like to make if you don't mind, which I will make after you merge. For example, it occurs to me that we could make yet another change to how we handle the timestamp field:

* If the value is of type `number` we just pass it through directly - no need to convert to Date object.
* If the value is of type `string` we try to get the browser to parse the date string. If the result is a valid Date object, we pass that along.
* If the value is of type `string` and we can't get the browser to create a valid Date object (such as Firefox when dealing with the `mm/dd/yy` format, then we will attempt to parse the string ourselves.

**Comment by breznak** [comment]
Thanks, merged. And :+1: on all of your suggestions for `timestamp`

**Comment by breznak** [comment]
My unsuccessful try on zoom in/out https://github.com/breznak/nupic/pull/14

**Comment by jefffohl** [comment]
@breznak I noticed that you removed "timestamp" from the `excludes` array used in the `generateFieldMap`function. We need to add this back in, or else we end up with 2 instances of "timestamp" in that array, because we have to unshift that item into the array at the end of the routine. This is because "timestamp" has to be at the beginning of the returned array.

**Comment by breznak** [comment]
> I noticed that you removed "timestamp" from the excludes array 

@jefffohl yes, sorry. It worked and I wasn't aware that it would double the column. 
You can just change to 
`var EXCLUDE_FIELDS = [TIMESTAMP];` to get it back. Will you do that, or should I?

**Comment by jefffohl** [comment]
@breznak - I can do it. Though, it depends on what your intentions are for `EXCLUDE_FIELDS`. If it is to make a list of fields that the user might want to exclude, then it might make more sense to hard-code it into the `generateFieldMap` function.

**Comment by breznak** [comment]
> If it is to make a list of fields that the user might want to exclude ...

@jefffohl I thought that was the meaning. If we hard-code it, will it still be possible to plot the "timestamp" column? (usually it's linear, although there can be jumps so it can be interesting being able to plot it too)

**Comment by jefffohl** [comment]
@breznak - In DyGraphs, the x-axis is always the timestamp. So, how would one plot the timestamp against the timestamp?

**Comment by breznak** [comment]
..ah, you got me :grinning:

**Comment by jefffohl** [comment]
New PR here: https://github.com/breznak/nupic/pull/15

**Comment by breznak** [comment]
@jefffohl merged, thank you!

**Comment by breznak** [comment]
Another thing I got stuck at https://github.com/breznak/nupic/pull/16 - a work to be able to highlight certain ranges when a field is over a threshold.

**Comment by breznak** [comment]
Small but needed doc in https://github.com/breznak/nupic/pull/17

**Comment by breznak** [comment]
some graph options in https://github.com/breznak/nupic/pull/19

**Comment by breznak** [comment]
@jefffohl in testing the UTC/local time, I've noticed an annoying usability problem: 
In the hotgym file, if you zoom to 1 week range, the daily trends are nicely visible, but the x-axis shows only days, so it's hard to get time for an exact point (one has to click it and see a pop-up window). So:

- [ ] UI: On higher zoom-out levels (~1 week) it is not possible to see time (fine-grained) on x-axis interactively. (I think best solution would be if the `timestamp` field would show up in the "fields div on the left", as interactive values are shown there.)

**Comment by breznak** [comment]
@jefffohl a couple of another ideas:

- `timestamp` we treat it as a special field as it's used as x-axis, to OPF it's just a normal field (only always present). Can we use `iteration` for x-axis instead and treat timestamp like any other field? (if it parses it will be included, no requirement on exact name/presence, no fallback hacks if parsing fails,... ) The disadvantage (?) would be linear iteration step (actually I'd think this is better - so in case there are jumps in timestamp we'll notice that, currently we cannot).

**Comment by breznak** [comment]
- `choice Input: OPF/CSV/NAB` to solve the "disadvantage" above, we could add a choice of input type (3 radio-choices right of the Render btn ?), that would simplify the code quite a lot:
  - for `OPF`: timestamp always called `timestamp`, is string, used as x-axis(?)
  - generic `CSV`: no requirements
  - `NAB` /optional/ could even start the webApp in an iframe below

**Comment by breznak** [comment]
@rhyolight looks like we're getting ready! If you have time, could you give this a shot? (and possibly improve the `visualization/Readme` ?)

**Comment by rhyolight** [comment]
@breznak @jefffohl Great work. I really like this. I tried it out and it works great. IMO you should merge with the current feature set, then email nupic-discuss with an explanation of how to use. See if any more feature requests come in. 

@breznak Go ahead and create a PR and I'll review. But I'm not sure I'll have time to update the readme but I'll try.

**Comment by jefffohl** [comment]
I can work on the README. Or is that something only the Flag Bearer is supposed to do?

**Comment by rhyolight** [comment]
@jefffohl Well there is no need to add to the main repo README. But the one in `scripts/visualization` should be created / updated.

**Comment by jefffohl** [comment]
@rhyolight - yes, I was referring to the README in `scripts/visualization`

**Comment by jefffohl** [comment]
@breznak do you feel that the current state of breznak/nupic/plot_results is what you would like to see for a first PR? If not, what features do you feel are needed? My understanding is that the outstanding features that in development are:

* Allow user to select timestamp field, or use the iteration step as the timestamp field
* Allow the user to select a threshold for each series. When this threshold is reached, the graph is highlighted.
* Show timestamp in labeled values (to the right of the graph)

Anything else?

**Comment by jefffohl** [comment]
@rhyolight - I have some tooling questions too. Should we be adding some unit tests here, and if so, do we have any existing Javascript testing frameworks in place within NuPIC? Also, if we are going to continue to build on this web app, I would like to start implementing some build tools to help with development (such as using `gulp` for build, `SaSS` for compiling CSS, `bower` for managing packages, etc.). This then forces the question - should this be inside NuPIC - or should it be its own project? (Since there is no dependency on NuPIC within this app)

**Comment by breznak** [comment]
Yes @jefffohl , I'm very satisfied with the current state of `plot_results`; it's more than what I've wanted at the beginning. 
Are you happy with the current state, OK for a PR? 

Those features in development are other enhancements, so not blocking the current state. I'll just reword them in my words: 

- improved 'generic CSV support' - timestamp would not be enforced anymore. Allow to toggle between time/step as x-axis. (Enhancement, https://github.com/breznak/nupic/pull/20 , almost complete-only needs the UI checkbox to function properly)
- highlighting subsections over the threshold; (and further development for AND-conditions) (New feature, https://github.com/breznak/nupic/pull/16 , early devel)
- for each selected (mouse hover) point show somewhere x-axis value (Enhancement) 
- allow to plot string/category data (New feature, in https://github.com/breznak/nupic/pull/20, 90% ready - need to fix the labels/columns count difference)

**Comment by rhyolight** [comment]
You are writing the first JavaScript I'm aware of in NuPIC. If you plan on
continuing, I would be more in favor of breaking out your plotting tool
into its own repository and adding tests there.

The reason is I don't really want to add a new JavaScript runtime to a
Python project just to run tests. I think it would be cleaner to have it in
its own repo, especially because of its generic functionality.

I say create a PR for now, but if this gets any bigger, move to another
repo in nupic-community.

---------
Matt Taylor
OS Community Flag-Bearer
Numenta

**Comment by breznak** [comment]
About the build tools - I can't judge how much they are needed, but I wouldn't be for extra dependencies to NuPIC, same for tests. This means something like `nupic.visualizations` repo would be nice(?) The nice thing now is you can fire off a browser and view a nupic results, so the new repo would have to be easily distributable - `pip` (?)

**Comment by jefffohl** [comment]
Sounds like we are agreed then. I will update the README, and we can make a PR [EDIT: @breznak already made a PR :)]. For further development, I can create a new repo, with a proper set of build tools, tests, etc.

**Comment by jefffohl** [comment]
I am not sure what the best way to distribute this would be, but `pip` seems like a strange fit - isn't that for Python packages only? For front-end apps like this, we would usually use something like Bower. Though,  most of the people in this community are probably not avid users of Bower. Of course, it should be pretty easy for people to just use git to clone the repo.

**Comment by breznak** [comment]
FTR..the work has moved to its stand-alone community repo here: https://github.com/nupic-community/nupic.visualizations
